### PR TITLE
Core 2.0 compatible unified extension API changes

### PR DESCRIPTION
**Pull request by spoorthipujariadobe**
## Description

1. Upgraded Java version , min, target and compile SDK versions
2. Use new `Log` service instead of `MobileCore.log`.
3. Replaced as `ExtensionListener` classes with `ExtensionEventListener` interface.
4. Replaced `MobileCore.dispatchEvent` with new method signature.
5. Replaced `MobileCore.dispatchEvent` in Extension class with `ExtensionApi.dispatch`
6. Modified public API class for new `registerExtensions` method and deprecated older `registerExtension` method.
7. Removed executor service usage in Extension class.
8. Added configuration dependency in `readyForEvent` and used new `getSharedState` API.
9. Replaced most casts with `DataReader`.
10. Added in checkstyle for code linting.
11. Fixed unit and functional tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java [hunk 1]
```diff
@@ -90,475 +69,548 @@ public void test_extensionVersion() {
...
-        assertEquals("com.adobe.eventType.optimize".toLowerCase(), event.getType());
-        assertEquals("com.adobe.eventSource.requestReset".toLowerCase(), event.getSource());
-        assertTrue(event.getEventData().isEmpty());
+        try (MockedStatic<MobileCore> mobileCoreMockedStatic = Mockito.mockStatic(MobileCore.class)) {
+            // test
+            Optimize.clearCachedPropositions();
+
+            //verify
+            final ArgumentCaptor<Event> eventCaptor = ArgumentCaptor.forClass(Event.class);
+            mobileCoreMockedStatic.verify(() -> MobileCore.dispatchEvent(eventCaptor.capture()), times(1));
```
Comment (swarna04) [inline code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java:607]: nit: `times(1)` is redundant here as `mobileCoreMockedStatic.verify(() -> MobileCore.dispatchEvent(eventCaptor.capture()));` verifies that certain behavior happened once. It is 
alias to verify(mock, times(1)). Check this in other tests as well.

File: code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java [hunk 2]
```diff
@@ -90,475 +69,548 @@ public void test_extensionVersion() {
...
+    }
 
-        // should be able to invoke the callback
-        // extensionErrorCallback.error(ExtensionError.UNEXPECTED_ERROR);
+    @Test
+    public void test_registerExtension_nullError() {
+        try (MockedStatic<MobileCore> mobileCoreMockedStatic = Mockito.mockStatic(MobileCore.class)) {
+            // setup
+            ArgumentCaptor<Class> extensionClassCaptor = ArgumentCaptor.forClass(Class.class);
+            ArgumentCaptor<ExtensionErrorCallback> callbackCaptor = ArgumentCaptor.forClass(
```
Comment (swarna04) [inline code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java:97]: Declare variables `final` e.g. final ArgumentCaptor<Class> extensionClassCaptor = ...

File: code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java [hunk 3]
```diff
@@ -90,475 +69,548 @@ public void test_extensionVersion() {
...
-        PowerMockito.verifyStatic(MobileCore.class, Mockito.times(1));
-        MobileCore.registerExtension(eq(OptimizeExtension.class), callbackCaptor.capture());
-
-        // verify the callback
-        ExtensionErrorCallback<ExtensionError> extensionErrorCallback = callbackCaptor.getValue();
-        assertNotNull("The extension error callback should not be null.", extensionErrorCallback);
+        try (MockedStatic<MobileCore> mobileCoreMockedStatic = Mockito.mockStatic(MobileCore.class)) {
+            // setup
+            ArgumentCaptor<Class> extensionClassCaptor = ArgumentCaptor.forClass(Class.class);
+            ArgumentCaptor<ExtensionErrorCallback> callbackCaptor = ArgumentCaptor.forClass(
```
Comment (swarna04) [inline code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java:75]: Declare variables final e.g. `final` ArgumentCaptor extensionClassCaptor = ...

File: code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java [hunk 4]
```diff
@@ -90,475 +69,548 @@ public void test_extensionVersion() {
...
+                    .when(() -> MobileCore.registerExtension(extensionClassCaptor.capture(), callbackCaptor.capture()))
+                    .thenReturn(true);
+            // test
+            Optimize.registerExtension();
+
+            // verify: happy
+            assertNotNull(callbackCaptor.getValue());
+            assertEquals(OptimizeExtension.class, extensionClassCaptor.getValue());
+            // verify: error callback was called
+            callbackCaptor.getValue().error(ExtensionError.UNEXPECTED_ERROR);
```
Comment (swarna04) [inline code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java:88]: nit: This `test_registerExtension()` should ideally be happy path test.  So test for `callbackCaptor.getValue().error(null);` in this test.

File: code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java [hunk 5]
```diff
@@ -90,475 +69,548 @@ public void test_extensionVersion() {
...
+            mobileCoreMockedStatic
+                    .when(() -> MobileCore.registerExtension(extensionClassCaptor.capture(), callbackCaptor.capture()))
+                    .thenReturn(true);
+            // test
+            Optimize.registerExtension();
+            // verify: happy
+            assertNotNull(callbackCaptor.getValue());
+            assertEquals(OptimizeExtension.class, extensionClassCaptor.getValue());
+
+            callbackCaptor.getValue().error(null);
```
Comment (swarna04) [inline code/optimize/src/test/java/com/adobe/marketing/mobile/optimize/OptimizeTests.java:109]: Suggest re-write
```
 @Test
    public void test_registerExtension_extensionError() {
        try (MockedStatic<MobileCore> mobileCoreMockedStatic = Mockito.mockStatic(MobileCore.class);
             MockedStatic<Log> logMockedStatic = Mockito.mockStatic(Log.class)) {
            // setup
            final ArgumentCaptor<Class> extensionClassCaptor = ArgumentCaptor.forClass(Class.class);
            final ArgumentCaptor<ExtensionErrorCallback> callbackCaptor = ArgumentCaptor.forClass(
                    ExtensionErrorCallback.class
            );
            mobileCoreMockedStatic
                    .when(() -> MobileCore.registerExtension(extensionClassCaptor.capture(), callbackCaptor.capture()))
                    .thenReturn(true);
            // test
            Target.registerExtension();

            // verify
            assertNotNull(callbackCaptor.getValue());
            assertEquals(TargetExtension.class, extensionClassCaptor.getValue());
            
            callbackCaptor.getValue().error(ExtensionError.UNEXPECTED_ERROR);
            logMockedStatic.verify(() -> Log.warning(anyString(), anyString(), anyString(), any()));
        }
    }
```

**Review by swarna04 — CHANGES_REQUESTED** [review]

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/DecisionScope.java [hunk 6]
```diff
@@ -22,12 +21,11 @@
 import java.util.Map;
 import java.util.Objects;
 
-import static com.adobe.marketing.mobile.optimize.OptimizeConstants.LOG_TAG;
-
 /**
  * {@code DecisionScope} class represents a scope used to fetch personalized offers from the Experience Edge network.
  */
 public class DecisionScope {
+    private static final String SELF_TAG = "DecisionScope";
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/DecisionScope.java:28]: The class name can get obfuscated if the app enables minify, so hard coding it. Following this convention from other core extensions https://github.com/adobe/aepsdk-core-android/blob/97622211024840a0bb3c1e99559535481b8a8365/code/android-signal-library/src/phone/java/com/adobe/marketing/mobile/Signal.java#L19

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Proposition.java [hunk 7]
```diff
@@ -118,37 +119,41 @@ public Map<String, Object> generateReferenceXdm() {
...
             if (OptimizeUtils.isNullOrEmpty(scope)) {
-                MobileCore.log(LoggingMode.DEBUG, LOG_TAG, "Cannot create Proposition object, provided data does not contain proposition scope.");
+                Log.debug(OptimizeConstants.LOG_TAG, SELF_TAG, "Cannot create Proposition object, provided data does not contain proposition scope.");
                 return null;
             }
 
-            final Map<String, Object> scopeDetails = (Map<String, Object>) data.get(OptimizeConstants.JsonKeys.PAYLOAD_SCOPEDETAILS);
+            final Map<String, Object> scopeDetails = DataReader.getTypedMap(Object.class, data, OptimizeConstants.JsonKeys.PAYLOAD_SCOPEDETAILS);
 
+            // todo
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Proposition.java:141]: `todo` was to replace casting with `DataReader` usage once nested generic types are supported. No sure if Core will add it. Will remove the `todo` once we decide on it

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.java [hunk 8]
```diff
@@ -408,54 +401,54 @@ public void error(final ExtensionError extensionError) {
      * @param data {@code Map<String, Object>} containing offer data.
      * @return {@code Offer} object or null.
      */
+    @SuppressWarnings("checkstyle:nestedifdepth")
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.java:404]: Added max depth as 3, will share internally but might be an extension specific setting

**Review by swarna04 — CHANGES_REQUESTED** [review]
Apologies! Adding additional feedback comments from a couple of files that I missed adding before.

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java [hunk 9]
```diff
@@ -196,19 +181,22 @@ public void call(final Event event) {
                     return;
                 }
 
+                // todo
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java:184]: `todo` was for potential `DataReader` usage. Will fix once Core team clarifies support for reading nested collections.

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java [hunk 10]
```diff
@@ -230,14 +218,17 @@ public void call(final Event event) {
...
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java:221]: Same as above

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java [hunk 11]
```diff
@@ -129,89 +161,78 @@ protected String getVersion() {
...
+        final Map<String, Object> configData = retrieveConfigurationSharedState(event);
+        if (OptimizeUtils.isNullOrEmpty(configData)) {
+            Log.debug(OptimizeConstants.LOG_TAG,
+                    SELF_TAG,
+                    "handleUpdatePropositions - Cannot process the update propositions request event, Configuration shared state is not available.");
+            return;
+        }
+
+        try {
+            // todo
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:176]: Placeholder for `DataReader`

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java [hunk 12]
```diff
@@ -223,63 +244,66 @@ public void error(final ExtensionError extensionError) {
...
+
+            // Verify the Edge response event handle
+            final String edgeEventHandleType = DataReader.getString(eventData, OptimizeConstants.Edge.EVENT_HANDLE);
+            if (!OptimizeConstants.Edge.EVENT_HANDLE_TYPE_PERSONALIZATION.equals(edgeEventHandleType)) {
+                Log.debug(OptimizeConstants.LOG_TAG, SELF_TAG,
+                        "handleEdgeResponse - Cannot process the Edge personalization:decisions event, event handle type is not personalization:decisions.");
+                return;
+            }
+
+            // todo
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:264]: Placeholder for DataReader

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java [hunk 13]
```diff
@@ -532,14 +499,25 @@ private List<String> retrieveValidDecisionScopes(final List<Map<String, Object>>
      *
      * @return {@link Event} instance.
      */
-    private Event createResponseEventWithError(final AdobeError error) {
+    private Event createResponseEventWithError(final AdobeError error, final Event event) {
         final Map<String, Object> eventData = new HashMap<>();
-        eventData.put(OptimizeConstants.EventDataKeys.RESPONSE_ERROR, error);
+        eventData.put(OptimizeConstants.EventDataKeys.RESPONSE_ERROR, error.getErrorName());
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:504]: `setEventData` is throwing serialization error with custom objects. I have informed this to core team. Added `todo`, will add a fix once they expose some methods to convert it

[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:505]: Discussed offline, will be passing `error.getErrorCode`. Using the error code to reconstruct `AdobeError` in Public API response callback

**Pull request merged**